### PR TITLE
Street distance calculation in backend standardized

### LINKED ISSUE #868: Completion on choropleth over 100%
This is on the production server. I am having a hard time understanding how this is the case, but I will think a bit more on it and look carefully at the code in the next day or two.

sidewalk-104-percent

I also found Takoma (West) to be at 101%.

### LINKED ISSUE #870: Neighborhood completion often won't reach 100%
While doing some testing to figure out why #868 is happening, I came across a situation where I audited an entire neighborhood, and the completion rate on the choropleth never reached 100%. I have already looked into the code a bit, and realized that I made a mistake in the code that updates the region_completion table after a street is audited. For streets that intersect multiple neighborhoods, I am only adding that distance to one of the neighborhoods that it intersects. Should be an easy fix.

### LINKED ISSUE #1078: Admin Interface: How Can Neighborhood Completion % Be Above 100?
How can neighborhood completion rates go beyond 100? Is this because we allow folks to audit neighborhoods that are already complete? Either way, probably there is a better way to visualize this :-P

image

### LINKED ISSUE #1158: Getting assigned neighborhoods that are already 100%. Stuck in endless loop.
Uh, oh--we expected this might happen (but hoped that it would not): it looks like users are experiencing issues with getting assigned neighborhoods that are already 100% completed (this is going to become an increasingly common problem due to the high completion rates we currently have).

The user nicely sent us this video over email:

https://drive.google.com/file/d/0B3uEwT_8YQHKSjlpZGFFQVRqc28/view?usp=sharing

The correct behavior is to assign a neighborhood with the *least* completion rate, right?

@misaugstad or @manaswis, can you investigate? I think this is related to https://github.com/ProjectSidewalk/SidewalkWebpage/issues/926, https://github.com/ProjectSidewalk/SidewalkWebpage/issues/795, and possibly https://github.com/ProjectSidewalk/SidewalkWebpage/issues/839

### PR DESCRIPTION
**Pull request by misaugstad**
Resolves #868 
Resolves #870 
Resolves #1078 
Resolves #1158 

Changes:
- replaced all instances of using JTS.transform to calculate street distance in Scala with calculation of distance within the query itself (required rewriting a couple queries in StreetEdgeTable.scala)
- fixed a line of code that could have placed a user in a region that they have already completed
- other minor fixes like the one just mentioned, that may or may not have caused problems elsewhere
- a bit of code refactoring, commenting, etc.

### CODE REVIEW

**Comment by misaugstad** [comment]
@manaswis and for the changes to take effect, just throw out all the records in the region_completion table. They will repopulate with the correct values.

File: app/controllers/AuditController.scala [hunk 1]
```diff
@@ -91,7 +91,8 @@ class AuditController @Inject() (implicit val env: Environment[User, SessionAuth
         UserCurrentRegionTable.assignNextEasyRegion(user.userId)
       
         var region: Option[NamedRegion] = RegionTable.selectTheCurrentNamedRegion(user.userId)
-        val task: NewTask = if (region.isDefined) AuditTaskTable.selectANewTaskInARegion(region.get.regionId, user) else AuditTaskTable.selectANewTask(user.username)
+        val task: NewTask = if (region.isDefined) AuditTaskTable.selectANewTaskInARegion(region.get.regionId, user)
+                            else AuditTaskTable.selectANewTask(user.username)
         Future.successful(Ok(views.html.audit("Project Sidewalk - Audit", Some(task), region, Some(user))))
```
Comment (misaugstad) [inline app/controllers/AuditController.scala]: reformatting b/c the line was really long

File: app/controllers/MissionController.scala [hunk 2]
```diff
@@ -33,7 +31,7 @@ class MissionController @Inject() (implicit val env: Environment[User, SessionAu
         // Mark the missions that should be completed.
         val regionId: Option[Int] = UserCurrentRegionTable.currentRegion(user.userId)
         if (regionId.isDefined) {
-          updatedUnmarkedCompletedMissionsAsCompleted(user.userId, regionId.get)
+          updateUnmarkedCompletedMissionsAsCompleted(user.userId, regionId.get)
         }
```
Comment (misaugstad) [inline app/controllers/MissionController.scala]: refactored to fix spelling

File: app/controllers/MissionController.scala [hunk 3]
```diff
@@ -108,18 +106,14 @@ class MissionController @Inject() (implicit val env: Environment[User, SessionAu
...
     */
-  def updatedUnmarkedCompletedMissionsAsCompleted(userId: UUID, regionId: Int): Unit = {
+  def updateUnmarkedCompletedMissionsAsCompleted(userId: UUID, regionId: Int): Unit = {
     val missions = MissionTable.selectIncompleteMissionsByAUser(userId, regionId)
-    val streets = StreetEdgeTable.selectStreetsAuditedByAUser(userId, regionId)
-    val CRSEpsg4326 = CRS.decode("epsg:4326")
-    val CRSEpsg26918 = CRS.decode("epsg:26918")
-    val transform = CRS.findMathTransform(CRSEpsg4326, CRSEpsg26918)
-    val completedDistance_m = streets.map(s => JTS.transform(s.geom, transform).getLength).sum
+    val completedDistance_m = StreetEdgeTable.getDistanceAudited(userId, regionId)
```
Comment (misaugstad) [inline app/controllers/MissionController.scala]: Switches from JTS.transform to calling a function in StreetEdgeTable that calculate the distance in the query.

File: app/models/audit/AuditTaskTable.scala [hunk 4]
```diff
@@ -97,6 +97,8 @@ object AuditTaskTable {
   val streetEdges = TableQuery[StreetEdgeTable]
   val users = TableQuery[UserTable]
 
+  val completedAuditTasks = auditTasks.filter(_.completed)
+
   case class AuditCountPerDay(date: String, count: Int)
```
Comment (misaugstad) [inline app/models/audit/AuditTaskTable.scala]: precomputed this here since it is used so often

File: app/models/audit/AuditTaskTable.scala [hunk 5]
```diff
@@ -301,7 +301,7 @@ object AuditTaskTable {
     * @return
     */
   def selectCompletedTasks(userId: UUID): List[AuditTask] = db.withSession { implicit session =>
-    auditTasks.filter(_.userId === userId.toString).list
+    completedAuditTasks.filter(_.userId === userId.toString).list
   }
```
Comment (misaugstad) [inline app/models/audit/AuditTaskTable.scala]: this was incorrectly looking at all audits, even incomplete ones

File: app/models/audit/AuditTaskTable.scala [hunk 6]
```diff
@@ -323,9 +323,10 @@ object AuditTaskTable {
 
     val completedTasks = for {
       u <- users.filter(_.username === username)
-      at <- auditTasks if at.userId === u.userId
+      at <- completedAuditTasks if at.userId === u.userId
     } yield (u.username.?, at.streetEdgeId.?)
```
Comment (misaugstad) [inline app/models/audit/AuditTaskTable.scala]: this also was not actually filtering out incomplete audits

File: app/models/audit/AuditTaskTable.scala [hunk 7]
```diff
@@ -449,7 +451,7 @@ object AuditTaskTable {
         StreetEdgeAssignmentCountTable.incrementAssignment(e.streetEdgeId)
         NewTask(e.streetEdgeId, e.geom, e.x1, e.y1, e.x2, e.y2, timestamp, completed=false)
       case _ =>
-        selectANewTask // The list is empty for whatever the reason. Probably the user has audited all the streets in the region
+        selectANewTask(userId) // The list is empty for whatever the reason. Probably the user has audited all the streets in the region
     }
```
Comment (misaugstad) [inline app/models/audit/AuditTaskTable.scala]: This method tries to get a new task in this region that the user has not audited, and if it finds none, it needs to get a new task from some other region. selectANewTask is an overloaded method, where no userId is taken in for anon users, but this method was erroneously calling the one for anon users. That version of the selectANewTask method does not filter out tasks that a user has already completed, so a user could have been assigned a task that they already did.

File: app/models/region/RegionCompletionTable.scala [hunk 8]
```diff
@@ -87,33 +83,29 @@ object RegionCompletionTable {
     * @return
     */
   def updateAuditedDistance(streetEdgeId: Int) = db.withTransaction { implicit session =>
-    val distToAdd: Float = streetEdgesWithoutDeleted.filter(_.streetEdgeId === streetEdgeId).groupBy(x => x).map(_._1.geom.transform(26918).length).list.head
-    val regionId: Int = streetEdgeNeighborhood.filter(_.streetEdgeId === streetEdgeId).groupBy(x => x).map(_._1.regionId).list.head
 
-    val q = for { regionCompletion <- regionCompletions if regionCompletion.regionId === regionId } yield regionCompletion
+    val distToAdd: Float = StreetEdgeTable.getStreetEdgeDistance(streetEdgeId)
```
Comment (misaugstad) [inline app/models/region/RegionCompletionTable.scala]: Replaced JTS.transform with that StreetEdgeTable func that does it in the query.

File: app/models/region/RegionCompletionTable.scala [hunk 9]
```diff
@@ -87,33 +83,29 @@ object RegionCompletionTable {
...
+      val q = for {regionCompletion <- regionCompletions if regionCompletion.regionId === regionId} yield regionCompletion
 
-    val updatedDist = q.firstOption match {
-      case Some(rC) => q.map(_.auditedDistance).update(rC.auditedDistance + distToAdd)
-      case None => -1
+      val updatedDist = q.firstOption match {
+        case Some(rC) => q.map(_.auditedDistance).update(rC.auditedDistance + distToAdd)
+        case None => -1
+      }
     }
```
Comment (misaugstad) [inline app/models/region/RegionCompletionTable.scala]: Before, the code was only updating the completion distance for a single region, even if a completed audit task intersected multiple regions. So I just modified it to update all of them.

File: app/models/region/RegionCompletionTable.scala [hunk 10]
```diff
@@ -87,33 +83,29 @@ object RegionCompletionTable {
...
 
       val neighborhoods = RegionTable.selectAllNamedNeighborhoods
       for (neighborhood <- neighborhoods) yield {
-        val streets: List[StreetEdge] = StreetEdgeTable.selectStreetsByARegionId(neighborhood.regionId)
-        val auditedStreets: List[StreetEdge] = StreetEdgeTable.selectAuditedStreetsByARegionId(neighborhood.regionId)
 
-        val auditedDistance = auditedStreets.map(s => JTS.transform(s.geom, transform).getLength).sum
-        val totalDistance = streets.map(s => JTS.transform(s.geom, transform).getLength).sum
+        val auditedDistance: Double = StreetEdgeTable.getDistanceAuditedInARegion(neighborhood.regionId).toDouble
+        val totalDistance: Double = StreetEdgeTable.getTotalDistanceOfARegion(neighborhood.regionId).toDouble
```
Comment (misaugstad) [inline app/models/region/RegionCompletionTable.scala]: Another case of replacing JTS.transform

File: app/models/street/StreetEdgeTable.scala [hunk 11]
```diff
@@ -288,31 +288,68 @@ object StreetEdgeTable {
...
+
+  /** Gets the query for a list of all street edges that the user has audited */
+  def selectAllStreetsAuditedByAUserQuery(userId: UUID) = db.withSession { implicit session =>
+
+    val auditedStreets = for {
+      (_streetEdges, _auditTasks) <- streetEdgesWithoutDeleted.innerJoin(completedAuditTasks).on(_.streetEdgeId === _.streetEdgeId)
+      if _auditTasks.userId === userId.toString
+    } yield _streetEdges
+    auditedStreets.groupBy(x => x).map(_._1) // does a select distinct
   }
```
Comment (misaugstad) [inline app/models/street/StreetEdgeTable.scala]: Rewrote the query using Scala/slick. But I also removed that inner join with the street_edge_region table b/c it wasn't doing anything. 

If you want a long-winded explanation of how it is not used, then here you go:
The street_edge_region table just has entries that pair every street edge with each neighborhood that it intersects. So most edges have 1 entry, and some have 2 (idk about more than that). And all street edges are in regions so its not like the inner join filters any of them out. It basically just adds a duplicate for each street edge that intersects multiple regions, then none of the columns in that table are used, and a select distinct is essentially done on the street_edge_id anyway.

File: app/models/street/StreetEdgeTable.scala [hunk 12]
```diff
@@ -288,31 +288,68 @@ object StreetEdgeTable {
...
-    val CRSEpsg26918 = CRS.decode("epsg:26918")
-    val transform = CRS.findMathTransform(CRSEpsg4326, CRSEpsg26918)
+  def getDistanceAudited(userId: UUID): Float = db.withSession { implicit session =>
 
-    val userStreets = selectAllStreetsAuditedByAUser(userId)
-    // get length of each street segment, sum the lengths, and convert from meters to miles
-    (userStreets.map(s => JTS.transform(s.geom, transform).getLength).sum * 0.000621371).toFloat
+    val dist = selectAllStreetsAuditedByAUserQuery(userId).groupBy(x => x).map(_._1.geom.transform(26918).length).list.sum
+    (dist * 0.000621371).toFloat // converts to miles
+  }
```
Comment (misaugstad) [inline app/models/street/StreetEdgeTable.scala]: Again replaces JTS.transform

File: app/models/street/StreetEdgeTable.scala [hunk 13]
```diff
@@ -288,31 +288,68 @@ object StreetEdgeTable {
...
+    // get the street edges from only this region
+    val auditedStreetsInRegion = for {
+      (_streetEdges, _streetEdgeRegions) <- selectAllStreetsAuditedByAUserQuery(userId).innerJoin(streetEdgeRegion).on(_.streetEdgeId === _.streetEdgeId)
+      if _streetEdgeRegions.regionId === region
+    } yield _streetEdges
+
+    // compute sum of lengths of the streets audited by the user in the region
+    val dist = auditedStreetsInRegion.groupBy(x => x).map(_._1.geom.transform(26918).length).list.sum
+    (dist * 0.000621371).toFloat // converts to miles
+  }
```
Comment (misaugstad) [inline app/models/street/StreetEdgeTable.scala]: This isn't used anywhere yet, but we are having problems with the total distance audited by users in missions right now, so I wrote this quickly and was printing out results to test things out. Will be used in the future.

File: app/models/street/StreetEdgeTable.scala [hunk 14]
```diff
@@ -288,31 +288,68 @@ object StreetEdgeTable {
...
+      if _streetEdgeRegions.regionId === regionId
+    } yield _streetEdges
+
+    // select distinct and sum the lengths of the streets
+    streetsInRegion.groupBy(x => x).map(_._1.geom.transform(26918).length).list.sum
+  }
+
+  /** Returns the distance of the given street edge */
+  def getStreetEdgeDistance(streetEdgeId: Int): Float = db.withSession { implicit session =>
+    streetEdgesWithoutDeleted.filter(_.streetEdgeId === streetEdgeId).groupBy(x => x).map(_._1.geom.transform(26918).length).list.head
```
Comment (misaugstad) [inline app/models/street/StreetEdgeTable.scala]: These three functions are the ones that replace JTS.transform in RegionCompletionTable.scala

File: app/models/street/StreetEdgeTable.scala [hunk 15]
```diff
@@ -333,7 +370,7 @@ object StreetEdgeTable {
     )
 
     try {
-      selectAuditedStreetsQuery(regionId).list
+      selectStreetsInARegionQuery(regionId).list
     } catch {
```
Comment (misaugstad) [inline app/models/street/StreetEdgeTable.scala]: These were both poorly-named variables, probably due to copy/pasting :)

**Comment by manaswis** [comment]
@misaugstad This PR has conflicts that needs resolution. Let me know when you do.

**Comment by manaswis** [comment]
@misaugstad I get this error when I load the load the application:

image

**Comment by manaswis** [comment]
@misaugstad I am not able to reach to 100% for Dupont Circle (246). I completed Dupont Circle West first during testing - 

image

It doesn't change even if I throw out region completion table records.

**Comment by manaswis** [comment]
I completed some other regions like Logan Circle/Shaw (East) and Logan Circle NorthEast and they reach 100%.

**Comment by misaugstad** [comment]
I was not able to reproduce :(

Did you say that you had dumped all the records from the regionCompletion table before starting to audit?

**Comment by manaswis** [comment]
I dumped it after auditing the two adjacent regions.


Best Regards,
Manaswi Saha
Ph.D. Student
Department of Computer Science
University of Maryland, College Park
http://cs.umd.edu/~manaswi/
Twitter - @manaswisaha <https://twitter.com/manaswisaha>

**Comment by misaugstad** [comment]
The issue you had mentioned about still not reaching 100% completion should be fixed once the most recent PR #1169 is merged. I am going to make one last pass through the code for this PR to make sure that it conforms to any changes we have made in the past 3 months, then it should be ready for testing/merging.

File: app/models/audit/AuditTaskTable.scala [hunk 16]
```diff
@@ -335,10 +335,11 @@ object AuditTaskTable {
     val timestamp: Timestamp = new Timestamp(Calendar.getInstance(TimeZone.getTimeZone("UTC")).getTime.getTime)
 
     val completedTasks = for {
-      u <- users.filter(_.username === user.toString)
-      at <- auditTasks if at.userId === u.userId
+      u <- users.filter(_.userId === user.toString)
```
Comment (misaugstad) [inline app/models/audit/AuditTaskTable.scala]: This was changed because the UUID toString method actually returns the userId, not the username.

**Comment by maddalihanumateja** [comment]
Put this message on the wrong PR: 
I'm going by what you said on slack. "The testing process would be the same as #1169 , except you wouldn’t have to truncate the table ever" . I restored the 10-16-2017 dump to test this. Georgetown had less than a mile remaining to audit so I chose to do missions there. I haven't seen a neighborhood completion message even after auditing around 3 miles. The landing page choropleth also doesn't seem to update for some reason. Could I be doing something incorrectly?

**Comment by manaswis** [comment]
@misaugstad said:

> Hmm, you probably needed to truncate the table once in the beginning possibly? If the table was out of sync to begin with, then it wouldn’t work. 

> Sorry for the imprecise instructions... so you get the dump, truncate the table, then do the testing as you said above

> Also, just making sure that you merged develop into this branch as well.

**Comment by maddalihanumateja** [comment]
@misaugstad could you resolve the conflict in this branch?

**Comment by maddalihanumateja** [comment]
This time I'd truncated the region completion table before starting auditing. I'm still being assigned 100% complete regions. Southwest employment are was assigned to me after I was done with a different neighborhood. It had 100% completion in the region completion table as well.  

image

I'm not sure if this is a pattern but I'm being assigned completed "difficult regions". Do you think this has anything to do with the part of the code that is assigning difficult regions to experienced users?

**Comment by maddalihanumateja** [comment]
*Seems to be working*. I had the following sequence of neighborhoods:
 - hill east  (west) (This was the only one that was 100% complete when I started working on it. I didn't get a neighborhood completion message here. Probably because the user had opted to continue working on the neighborhood)
 - monumental core (No neighborhood completion message. The mission completion popup showed an "audit a different neighborhood" button directly )
 - navy yard (No neighborhood completion message. The mission completion popup showed an "audit a different neighborhood" button directly )
 - southwest waterfront
 - benning 
 - manor park

All the neighborhoods that I was assigned to with the exception of the initial one were less than 100% complete according to the region completion table.

**Pull request merged**